guidance/library/_json.py

```diff
@@ -359,16 +360,24 @@ def _gen_json_object(
     required: Sequence[str],
     definitions: Mapping[str, Callable[[], GrammarFunction]],
 ):
-    if any(k not in properties for k in required):
-        raise ValueError(f"Required properties not in properties: {set(required) - set(properties)}")
-
-    grammars = tuple(f'"{name}":' + _gen_json(json_schema=schema, definitions=definitions) for name, schema in properties.items())
-    required_items = tuple(name in required for name in properties)
+    # "required" keys will be validated against "properties" if they're present, otherwise against "additionalProperties".
+    # If "additionalProperties" is False, then required keys must be in "properties".
+    if any(k not in properties for k in required) and additional_properties is False:
+        raise ValueError(
```

Comment:
Not strictly necessary, as generating with a `False` schema will raise an exception later. But this lets us give a more transparent error.